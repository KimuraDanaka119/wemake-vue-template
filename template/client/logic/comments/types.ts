import { RawCommentType } from '~/logic/comments/models'

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number
}

export interface CommentPayloadType {
  commentId: number
  delta: number
}
