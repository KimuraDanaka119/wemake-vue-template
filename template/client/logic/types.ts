import { CommentType } from '~/logic/comments/types'

// TODO: document

export interface StateType {
  // Module:
  comments: {
    // Module's state:
    comments: CommentType[]
  }
}
