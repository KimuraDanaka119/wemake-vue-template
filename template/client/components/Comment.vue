<template>
  <div
    class="comment__component"
  >
    <p class="comment-author">{{ comment.email }}</p>
    <p class="comment-body">{{ comment.body }}</p>

    <div class="comment-rating">
      <button
        @click="changeRating(comment.id, 1)"
        class="comment-rating-up">+</button>

      <span class="comment-rating-value">{{ comment.rating }}</span>

      <button
        @click="changeRating(comment.id, -1)"
        class="comment-rating-down">-</button>
    </div>
  </div>
</template>

<script>
// @flow

import * as mutationTypes from '~/types/mutations'

export default {
  props: {
    comment: {
      type: Object,
      required: true
    }
  },

  methods: {
    changeRating (commentId: number, delta: number) {
      this.$store.commit(mutationTypes.UPDATE_RATING, { commentId, delta })
    }
  }
}
</script>

<style lang="scss" scoped>
.comment__component {
  background: white;
  max-width: 15.6rem;
  border: 0.12rem solid;
  padding: 0.74rem;
  margin: 0.5rem;
  position: relative;

  .comment-author {
    padding-bottom: 0.74rem;
    font-weight: bold;
    border-bottom: 0.12rem solid;
  }

  .comment-rating {
    position: absolute;
    bottom: 0.3rem;
    left: 50%;
    transform: translateX(-50%);

    button {
      margin: 0 0.5rem;
      border: none;
      outline: none;
      cursor: pointer;
      width: 2rem;

      &.comment-rating-up {
        background: #a2dca2;
      }

      &.comment-rating-down {
        background: #efbcd1;
      }
    }
  }
}
</style>
