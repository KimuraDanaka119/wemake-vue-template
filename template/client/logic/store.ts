import { Module } from 'vuex-simple'

// TODO: document

import CommentsModule from '~/logic/comments/module'

export default class TypedStore {
  @Module()
  public comments = new CommentsModule()
}
