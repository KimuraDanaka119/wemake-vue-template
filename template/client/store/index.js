// @flow

import Vue from 'vue'

import comments from '~/logics/api/comments'
import * as mutationTypes from '~/types/mutations'
import type {
  CommentType,
  CommentPayloadType,
  StateType,
  RawCommentType
} from '~/types'

function state (): StateType {
  return {
    comments: null
  }
}

const getters = {
  hasComments (state: StateType): boolean {
    return Boolean(state.comments && state.comments.length > 0)
  }
}

const mutations = {
  [mutationTypes.SET_COMMENTS]: (
    state: StateType, comments: Array<RawCommentType>
  ) => {
    const updatedComments: Array<CommentType> = []

    for (const comment of comments.slice(0, 10)) {
      const newOne = {
        ...comment,
        rating: 0
      }
      updatedComments.push(newOne)
    }

    state.comments = updatedComments
  },

  [mutationTypes.UPDATE_RATING]: (
    state: StateType, { commentId, delta }: CommentPayloadType
  ) => {
    if (!state.comments) return

    const commentIndex = state.comments.findIndex((c: CommentType) => {
      return c.id === commentId
    })

    if (!state.comments || !state.comments[commentIndex]) return

    state.comments[commentIndex].rating += delta
  }
}

const actions = {
  async fetchComments ({ commit }: any, app: Vue) {
    const data = await comments.fetchComments(app)
    commit(mutationTypes.SET_COMMENTS, data)

    return data
  }
}

export {
  state,
  getters,
  mutations,
  actions
}
