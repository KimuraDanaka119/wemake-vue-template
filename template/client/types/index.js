// @flow

export type RawComment = {
  id: number,
  body: string,
  email: string,
};

// We add rating only on the client (for demo purposes)
export type Comment = RawComment & {
  rating: number
};

export type State = {
  comments: ?Array<Comment>
};
