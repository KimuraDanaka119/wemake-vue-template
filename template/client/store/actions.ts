import { ActionContext, Store } from 'vuex'

import comments from '~/logics/api/comments'
import * as mutationTypes from '~/store/types'
import { StateType } from '~/types'

const actions = {
  async fetchComments (
    this: Store<StateType>,
    { commit }: ActionContext<StateType, StateType>,
  ) {
    const commentsList = await comments.fetchComments(this.$axios)
    commit(mutationTypes.SET_COMMENTS, commentsList)
    return commentsList
  },
}

export default actions
