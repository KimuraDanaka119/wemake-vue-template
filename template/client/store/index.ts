// We use default Nuxt Module-based store,
// read more about it here:
// https://nuxtjs.org/guide/vuex-store

// TODO: document

import Vue from 'vue'
import Vuex, { Store, Plugin } from 'vuex'
import { createVuexStore } from 'vuex-simple'
import { Container } from 'vue-typedi'

import tokens from '~/logic/tokens'
import TypedStore from '~/logic/store'
import { StateType } from '~/logic/types'

Vue.use(Vuex)

export default function store (
  ssrContext,
  { plugins = [] }: { plugins: Plugin<StateType>[] },
): Store<StateType> {
  if (!Container.has(tokens.AXIOS)) {
    Container.set(tokens.AXIOS, (): void => {})
  }

  const typedStore = new TypedStore()

  // Registering DI container items:
  Container.set(tokens.STORE, typedStore)

  return createVuexStore(typedStore, {
    plugins,
  })
}
