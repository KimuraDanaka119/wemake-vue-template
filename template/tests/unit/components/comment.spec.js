import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'

import Comment from '~/components/Comment'
import { state, mutations } from '~/store'
import * as types from '~/types/mutations'

import { createStore } from '../../utils/store'

const localVue = createLocalVue()
localVue.use(Vuex)

const comment = {
  id: 1,
  email: 'test@mail.com',
  body: 'some text',
  rating: 0
}

const propsData = { comment }

describe('Comment component unit tests', () => {
  let store

  beforeEach(() => {
    store = createStore({ state, mutations }, {
      comments: [comment]
    })
  })

  test('should have two buttons', () => {
    const wrapper = mount(Comment, { store, localVue, propsData })

    expect(wrapper.findAll('button')).toHaveLength(2)
  })

  test('should have correct values', () => {
    const wrapper = mount(Comment, { store, localVue, propsData })

    expect(wrapper.find('.comment-author').text().trim()).toEqual(comment.email)
    expect(wrapper.find('.comment-body').text().trim()).toEqual(comment.body)
    expect(wrapper.find('.comment-rating-value').text().trim()).toEqual(
      comment.rating.toString()
    )
  })

  test('should increment rating', () => {
    const wrapper = mount(Comment, { store, localVue, propsData })

    wrapper.vm.$store.commit(types.UPDATE_RATING, {
      commentId: comment.id,
      delta: 1
    })
    expect(wrapper.vm.$store.state.comments[0].rating).toEqual(1)
  })

  test('should decrement rating', () => {
    const wrapper = mount(Comment, { store, localVue, propsData })

    wrapper.vm.$store.commit(types.UPDATE_RATING, {
      commentId: comment.id,
      delta: -1
    })
    expect(wrapper.vm.$store.state.comments[0].rating).toEqual(-1)
  })

  test('should match the snapshot', () => {
    const wrapper = mount(Comment, { store, localVue, propsData })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
