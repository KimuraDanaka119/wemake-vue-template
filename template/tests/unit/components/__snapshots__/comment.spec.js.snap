// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit tests for Comment component should match the snapshot 1`] = `
"<div class=\\"comment__component\\"><p class=\\"comment-author\\">test@mail.com</p> <p class=\\"comment-body\\">some text</p> <div class=\\"comment-rating\\"><button class=\\"comment-rating-up\\">
      +
    </button> <span class=\\"comment-rating-value\\">0</span> <button class=\\"comment-rating-down\\">
      -
    </button></div></div>"
`;
